05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb

```diff
@@ -0,0 +1,594 @@
+      "    (6): Linear(in_features=4096, out_features=1000)\n",
+      "  )\n",
+      ")\n"
+     ]
+    }
+   ],
+   "source": [
+    "from torchvision import models\n",
+    "model = models.vgg19(pretrained=True)\n",
+    "print(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The model has two parts. \n",
+    "\n",
+    "model.features is the convolutional part, containing only convolutional layers, ReLU and max pooling layers.\n",
+    "\n",
+    "\n",
+    "model.classifier contains several fully connected and ReLU layers and does the final classification on features extracted from model.features. \n",
+    "\n",
+    "We will be focusing on model.features. Lets look at one of the convolutional layers:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 36,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Conv2d (64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))\n"
+     ]
+    }
+   ],
+   "source": [
+    "layer = model.features._modules['2']\n",
+    "print(layer)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This layer performs 64 different convolutions (the output channels), \n",
+    "\n",
+    "where each convolution has a 3x3 spatial size and is done on an image with 64 channels (the input channels).\n",
+    "\n",
+    "If we call our convolutional kernel K, the input image X and the layer bias b, then the output from the layer, V will be:\n",
+    "\n",
+    "$$ V(x, y, t) = \\sum_i \\sum_j \\sum_sK(i, j, s, t)X(x-i, x-j, s) + b(t) $$ \n",
+    "\n",
+    "\n",
+    "The weights of the layer are stored in a 4 dimensional tensor:\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "torch.Size([64, 64, 3, 3])\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(layer.weight.size())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will use TensorLy for applying tensor decompositions on this 4 dimensional weight tensor.\n",
+    "\n",
+    "\n",
+    "## CP Decomposition on convolutional layers\n",
+    "\n",
+    "The idea here is from [1412.6553 Speeding-up Convolutional Neural Networks Using Fine-tuned CP-Decomposition](https://arxiv.org/abs/1412.6553)\n",
+    "\n",
+    "We can approximate our 4 dimensional convolutional kernel with a CP decomposition of rank R:\n",
+    "\n",
+    "$ K(i, j, s, t) = \\sum_{r=1}^R K^x_r(i)K^y_r(j)K^s_r(s)K^t_r(t) $.\n",
+    "\n",
+    "Plugging this into the formula for the convolutional layer output form above:\n",
+    "\n",
+    "$ V(x, y, t) = \\sum_r\\sum_i \\sum_j \\sum_sK^x_r(i)K^y_r(i)K^s_r(s)K^t_r(t)X(x-i, y-j, s) $\n",
+    "$ = \\sum_rK^t_r(t) \\sum_i \\sum_j K^x_r(i)K^y_r(j)\\sum_sK^s_r(s)X(x-i, y-j, s) $ \n",
```

Comment:
I guess later on we can add some images and equivalent formulation to make the whole formula more human friendly :)